src/pages/homeCollection/kitAssembly.js

```diff
@@ -1,21 +1,391 @@
+            <td>${kitData[i].collectionCardId}</td>
+        </tr>`;
+
+    // Update the last row number
+    lastRowNumber = i + 1;
+    // // If the current iteration is the last item and matches length of last row variable, add an extra row
+    if (lastRowNumber === kitData.length) {
+      extraRow = `
+        <tr class="new-row">      
+          <th scope="row">${lastRowNumber + 1}</th>
+          <td>
+              <input id="input-usps" type="string" autocomplete="off" style="width:80%"/>
+          </td>
+          <td>
+              <input id="input-supply-kit" type="string" autocomplete="off" style="width:80%"/>
+          </td>
+          <td>
+              <input id="input-specimen-kit" type="string" autocomplete="off" style="width:80%"/>
+          </td>
+          <td>
+              <input id="input-collection-cup" type="string" autocomplete="off" style="width:80%"/>
+          </td>
+          <td>
+              <input id="input-collection-card" type="string" autocomplete="off" style="width:80%"/>
+          </td>
+      </tr>
+      `;
+      tableRow += extraRow;
+    }
+
+    tableBody.innerHTML = tableRow;
+  }
+};
+
+const kitAssemblyPageButtons = () => {
+  const contentBody = document.getElementById("contentBody");
+  let buttonContainerTemplate = "";
+
+  buttonContainerTemplate += `
+        <div class="kit-assembly-button-container d-flex justify-content-around" style="margin:8rem 0;">
+          <button id="kit-assembly-cancel-button" type="button" class="btn btn-outline-secondary" style=" width:13rem; height:3rem; border-radius:15px">Cancel</button>
+
+          <button id="kit-assembly-save-button" type="submit" class="btn btn-success" style="width:13rem;height:3rem; border-radius:15px">Save</button>
+        </div> 
+    `;
+  contentBody.innerHTML += buttonContainerTemplate;
+};
+
+const saveItem = async (
+  tableBody,
+  inputUsps,
+  inputSupplyKit,
+  inputSpecimenKit,
+  inputCollectionCup,
+  inputCollectionCard,
+  inputElements
+) => {
+  const saveButton = document.getElementById("kit-assembly-save-button");
+  let tableNumRows = tableBody.rows.length;
+  saveButton.addEventListener("click", (e) => {
+    tableNumRows++;
+
+    e.preventDefault();
+
+    // Target Last row and the last row's children elements
+    // Remove whitespace if any on input fields
+    jsonSaveBody.collectionCardId = inputCollectionCard.value.trim();
+    jsonSaveBody.supplyKitId = inputSupplyKit.value.trim();
+    jsonSaveBody.collectionCupId = inputCollectionCup.value.trim();
+    jsonSaveBody.specimenKitId = inputSpecimenKit.value.trim();
+    jsonSaveBody.uspsTrackingNumber = inputUsps.value.trim();
+
+    // PREVENTS USER FROM SUBMITTING INCOMPLETE INPUT FIELD ROW
+    for (const key in jsonSaveBody) {
+      if (!jsonSaveBody[key]) {
+        alert("One or more inputs are empty!");
+        return;
+      }
+    }
+
+    // ADD DATA to TABLE
+    addRow(jsonSaveBody);
+    function addRow(jsonSaveBody) {
```

Comment:
did you define a function & make a function call within the same function (twice)?
I'm not sure if this is a good practice? Can you put the function definitions outside saveItem()?
If you think this is a good practice, lmk?

Reply:
This makes sense. Correct me if I am wrong, having the function declarations addRow and clearRowInputs inside saveItem function would be a waste of resources. The user would be clicking on the submit button repeatedly and the two function declarations would be declared multiple times. But from your suggestion and the optimal solution, by having both addRow and clearRowInputs on the outer scope the function is declared only once within the scope of the module and not multiple times. The addRow and clearRowInputs functions are invoked/called multiple times with each button press that calls the saveItem function.